src/lib.rs

```diff
@@ -160,22 +159,30 @@ impl<T: FileLoader, U> VfsInternal<T, U> {
         }
     }
 
-    fn on_changes(&self, changes: &[Change]) -> Result<(), Error> {
-        for (file_name, changes) in coalesce_changes(changes) {
+    fn on_changes(&self, changes: &Changes) -> Result<(), Error> {
+        for (file_name, change) in changes {
             let path = Path::new(file_name);
-            {
-                let mut files = self.files.lock().unwrap();
-                if let Some(file) = files.get_mut(Path::new(path)) {
-                    file.make_change(&changes)?;
-                    continue;
+            let mut files = self.files.lock().unwrap();
+            match *change {
+                FileChange::Add { ref text } => {
+                    files.insert(path.to_path_buf(), File::from_text(text.clone()));
+                }
+                FileChange::Change { ref edits } => {
+                    let mut file = match files.entry(path.to_path_buf()) {
+                        hash_map::Entry::Occupied(o) => o.into_mut(),
+                        // FIXME: this is subtly broken, because we can't guarantee that the edits
+                        // are intended to be applied to the version of the file we read from disk.
+                        // That is, the on disk contents might have changed after the edit request.
+                        // The proper solution is to error here and to require that all files are
+                        // added first.
```

Comment:
Hmm, this is an interesting point. I think there is something to be said for having the VFS be able to load data from disk itself, rather than just work on supplied text. In particular, that makes it easier for it to work with timestamps on the files and thus make correct writes. However, in the IDE case, the IDE is totally responsible for reading and writing the files, so it seems right to error out here like you suggest. Perhaps the VFS should be configured in one mode or the other when it is started up.